### ISSUE TITLE
Add support for markdown reference-style link

### ISSUE THREAD

**Issue by JasonMFry** [vimwiki/vimwiki#925]
I'd like support for markdown reference-style links. I don't like using inline links b/c it makes the line length very long, and if I wrap the line it almost always breaks up the markdown link, which I also don't like. Support for reference-style links would address both of those issues.

This is what "support" means to me:

1. Pressing `enter` on a word creates a reference-style link
2. Pressing `enter` on a reference-style link takes you to the link

I've just started using vimwiki in the past week so maybe there are more things that "support" should mean but that's all I can think of.

I'm willing to put up a PR for this, especially if someone points me in the right directions to get started, but it may very well take months for me to finish b/c I don't have much free time.

**Comment by tinmarino** [comment]
#936 
vars.vim
spagethi code warning

**Comment by patdavid** [comment]
I'd be willing to help however I can for this also, although I might require some hand-holding at first.  I can promise I will do my best to learn, though.

Reference-style links are important enough for me to at least try to help. :)

**Comment by tinmarino** [comment]
Hi,

Good initiative @patdavid read vimwiki#base#follow_link. I am glad to see some motivation to contribute to vimwiki and try to make a code documentation whenever possible. @JasonMFry have a look at #1005. It may take 5h of reading code, to understand how plugin and ftplugin are adding commands linking to autoload/base which uses all other files. u.vim sandars for util and vars.vim is to keep all user configuration variable in one place as explained [here](https://github.com/vimwiki/vimwiki/blob/dev/doc/design_notes.md#scoped-variable).

**Comment by tinmarino** [comment]
Little doc for the brave coder.

[Link reference definition](https://github.github.com/gfm/#link-reference-definitions) convert:
```markdown
- [2020-07-11-three-weeks-in]

[2020-07-11-three-weeks-in]: blog/2020-07-11-three-weeks-in.md "Three Weeks In"
```
in
```html
<ul>
<li><a href="blog/2020-07-11-three-weeks-in.md" title="Three Weeks In">2020-07-11-three-weeks-in</a></li>
</ul>
```

**Comment by tinmarino** [comment]
@patdavid  if you make the PR before November you gain 1/4 of Hacktoberfest T-shirt Yeepi

**Comment by apruden2008** [comment]
@tinmarino I started using Vim this year and love it. Recently discovered this plugin and think it's wonderful for many applications. Not sure if anyone is still working on this but I'd love to give it a go.

**Comment by tinmarino** [comment]
Hello @apruden2008,
Glad to hear from you. Just give it a go. @patdavid would be so glad to have his golden feature done. 

In absence of any return after a month, I may reafect it to another person. Don't feel stress or pushed thought, I said "return" not code, just a sign of life if someone pings here.

Feel free to ask but you may not receive answers because there are so many tickets, sometimes I loose myself.

**Comment by apruden2008** [comment]
Thanks. Quick clarification @patdavid on the implementation. Pressing <CR> right now creates an embedded link in vimwiki. Should there be some kind of flag the user can set to differentiate between embedded and reference style? Or should this reassign <CR> to only create reference style links?

I personally prefer the option, so that's how I'm going to start implementing for now.

**Comment by tinmarino** [comment]
Opcion: link_type = reference or web 
That will enable us to add the `[[link]]` type later

**Comment by apruden2008** [comment]
The regex patterns in here are dense. Any tips on how to decipher or how to get up to speed? I know some basic regex but this is very hard to parse: 

\[\[[^\\\]]\{-}\%(|[^\\\]]\{-}\)\?\]\]\|[\]\[]\@<!\[[^\\\[\]]\{-}\]\[\][\]\[]\@ !\|[\]\[]\@<!\[[^\\\[\]]\{-}\]\[[^\\\[\]]\{-}\][\]\[]\@!

**Comment by tinmarino** [comment]
@apruden2008 good question. I think there is no way.
What I do is copy it in a draft and separe whaa is separable with spaces, new lines and comments. But each one of us has to do it again an again.
The best dev can do it to create a regex factory file and parametrize, configure them.
In adition, there is often a match and search pattern that are almost equivalent => code duplication.
Feel free to make some PR with only comments and refactorings.
Sorry ..

**Comment by apruden2008** [comment]
No worries. I'm here to contribute to an open-source project that I like and use, not to earn a free t-shirt. Moreover, it's good practice. But I understand if you want to prioritize it and pass to a regex pro. But until then, I'll keep hacking at it and see what I can come up with.

**Comment by tinmarino** [comment]
> But I understand if you want to prioritize it and pass to a regex pro

No hurry, take your time, we have 140 open ticket so there is work for
every one.

**Comment by patdavid** [comment]
Sorry, was afk for the weekend.  I really have the best intentions to try and help out, but am stretched a bit thin at the moment so haven't had a chance to.  I apologize.

**Comment by themarcel** [comment]
I've set up an LspAttach autocmd to trigger your plugin `LinkConvertAll` command on `BufEnter` of the LSP hover buffer, allowing to edit links in a reference-style format automatically.

```lua
vim.api.nvim_create_autocmd("LspAttach", {
	callback = function(args)
		local client = vim.lsp.get_client_by_id(args.data.client_id)
		if not client then
			return
		end

		local original_hover = vim.lsp.buf.hover

		-- Override the hover function
		vim.lsp.buf.hover = function()
			original_hover()

			vim.api.nvim_create_autocmd("BufEnter", {
				once = true,
				callback = function()
					local should_format = vim.bo.filetype == "markdown"
						and vim.bo.buftype == "nofile"

					if should_format then
						vim.cmd.set "modifiable"
						vim.cmd "LinkConvertAll"
					end
				end,
			})
		end
	end,
})
```

Example of the result:
![demo](https://github.com/user-attachments/assets/94f9fbdf-4367-403b-a2ed-4672fe607116)

**Comment by qadzek** [comment]
@marcelarie, thanks for sharing your snippet and screencast! It's great to see how users use the plugin in different ways.

If you think your snippet could benefit more users, feel free to add it to the plugin's wiki: https://github.com/qadzek/link.vim/wiki

**Comment by themarcel** [comment]
@qadzek I will add it when I find some time, but it might be a nice addition to the plugin via a global boolean value to activate it.
